ingest-router/src/relay_project_config_handler.rs

```diff
@@ -0,0 +1,1013 @@
+    /// if the task times out or panics.
+    fn spawn_upstream_tasks(
+        &self,
+        split_requests: Vec<(Upstream, RelayProjectConfigsRequest)>,
+        base_request: &Request<()>,
+    ) -> Result<Vec<TaskWithKeys>, IngestRouterError> {
+        let mut tasks = Vec::new();
+
+        for (upstream, split_request) in split_requests {
+            // Track public keys for this upstream (needed if request fails)
+            let public_keys = split_request.public_keys.clone();
+
+            // Serialize the split request body
+            let request_body = split_request.to_bytes().map_err(|e| {
+                IngestRouterError::InternalError(format!("Failed to serialize request: {e}"))
+            })?;
+
+            // Build request with headers from original request
+            let mut req_builder = Request::builder()
+                .method(base_request.method())
+                .uri(base_request.uri().clone())
+                .version(base_request.version());
+
+            for (name, value) in base_request.headers() {
+                req_builder = req_builder.header(name, value);
+            }
+
+            let request = req_builder
+                .body(Full::new(request_body))
+                .expect("Failed to build request");
+
+            let client = self.client.clone();
+            let sentry_url = upstream.sentry_url.clone();
+
+            // Clone keys so we can track them even if task times out
+            let public_keys_for_task = public_keys.clone();
+
+            // Spawn a task for each upstream request
+            let task = tokio::spawn(async move {
+                let sentry_url_str = sentry_url.to_string();
+                let result: Result<Response<Bytes>, IngestRouterError> =
+                    send_to_upstream(&client, &sentry_url, request, 30).await;
+
+                UpstreamTaskResult {
+                    sentry_url: sentry_url_str,
+                    public_keys: public_keys_for_task,
+                    result,
+                }
+            });
+
+            // Store both task handle and keys (keys needed if task times out)
+            tasks.push(TaskWithKeys {
+                handle: task,
+                public_keys,
+            });
+        }
+
+        Ok(tasks)
+    }
+
+    /// Collect results from all spawned tasks and merge them.
+    ///
+    /// Handles timeouts, task panics, and HTTP failures gracefully.
+    /// Failed keys are added to the pending array for retry.
+    async fn collect_task_results(&self, tasks: Vec<TaskWithKeys>) -> MergedResults {
+        let mut merged_configs = HashMap::new();
+        let mut merged_pending = Vec::new();
+        let mut merged_extra = HashMap::new();
+
+        for task_with_keys in tasks {
+            let TaskWithKeys {
+                mut handle,
+                public_keys,
+            } = task_with_keys;
+            let task_result = timeout(Duration::from_secs(30), &mut handle).await;
```

Comment:
I sort of wonder if we need more than 1 timeout value for various scenarios. The specific cases i'm thinking about are:
- i think a really long value like 30 secs is fine if we have no responses from any cell -- without any data relay is making no progress anyway so it's not any worse to wait longer for a response
- If 1 cell returns in say 200 milliseconds, and the other cell is down, waiting for the full 30 seconds for the other cell may be a serious problem. Since the requests are chunked this become a very large amount of 30 second waits back to back. If we returned sooner, the cell that was functional would be able to make progress much sooner. I'm worried that such a long wait for an unresponsive cell may risk effectively taking down all other cells as well.

Reply:
> I sort of wonder if we need more than 1 timeout value for various scenarios.

Having multiple configuration points for different timeouts seems wise. Waiting here would be sequential which isn't ideal. Perhaps we could combine a [join_set](https://docs.rs/tokio/latest/tokio/task/struct.JoinSet.html) to wait on the task futures concurrently, and a timeout so that we get as many responses as possible within the 30s timeout?